src/foam/web/URLState.js

```diff
@@ -0,0 +1,236 @@
+          creationContext: this.__subContext__
+        });
+      }
+    },
+    {
+      name: 'bindingsMap_',
+      documentation: 'Map of {<key>: foam.core.Slot} comprising state.',
+      factory: function() { return {}; }
+    },
+    {
+      name: 'subMap_',
+      documentation: `Map of {<key>: <event subscription>} parallel to
+          bindingsMap_.`,
+      factory: function() { return {}; }
+    },
+    {
+      name: 'unboundMap_',
+      documentation: `Map of {<key>: <deserialized value>} that contains
+          bindings loaded from hash, but not bound to a foam.core.Slot.`,
+      factory: function() { return {}; }
+    },
+    {
+      class: 'String',
+      documentation: `"path" part of hash; hash is:
+          "#<path>?<key1>=<value1>&<key2>=<value2>..."`,
+      name: 'path_',
+      postSet: function(old, nu) {
+        if ( old !== nu ) this.onStateChange();
+      }
+    }
+  ],
+
+  methods: [
+    function init() {
+      this.SUPER();
+      this.hashToState_();
+      this.window.addEventListener('hashchange', this.onHashChange);
+    },
+    function getPath() { return this.path_; },
+    function setPath(path) { return this.path_ = path; },
+    function getSlot(name) {
+      return this.bindingsMap_[name] || null;
+    },
+    function addBinding(name, slot) {
+      if ( this.bindingsMap_.hasOwnProperty(name) ) {
+        this.warn('Overwriting URLState:', name, this.bindingsMap_[name],
+                  'with', slot);
+        this.subMap_[name].detach();
+      }
+      if ( this.unboundMap_.hasOwnProperty(name) ) {
+        slot.set(this.unboundMap_[name]);
+        delete this.unboundMap_[name];
+      }
+      this.bindingsMap_[name] = slot;
+      this.subMap_[name] = slot.sub(this.onStateChange);
+      this.onStateChange();
+    },
+    function removeBinding(name) {
+      if ( ! this.bindingsMap_.hasOwnProperty(name) ) return;
+      this.subMap_[name].detach();
+      delete this.subMap_[name];
+      delete this.bindingsMap_[name];
+      this.onStateChange();
+    },
+    function clearBindings() {
+      var subMap = this.subMap_;
+      for ( var key in subMap ) {
+        if ( ! subMap.hasOwnProperty(key) ) continue;
+        subMap[key].detach();
+      }
+      this.subMap_ = {};
+      this.bindingsMap_ = {};
+      this.unboundMap_ = {};
+      this.onStateChange();
+    },
+    function stateToHash_() {
+      var hash = '#' + this.path_ + '?';
+      var bindingsMap = this.bindingsMap_;
+      var first = true;
+      var value;
+      for ( var key in bindingsMap ) {
+        if ( ! bindingsMap.hasOwnProperty(key) ) continue;
+
+        if ( ! first ) hash += '&';
+        first = false;
+
+        value = this.serializer.stringify(bindingsMap[key].get());
+        hash += key + '=' + this.window.encodeURIComponent(value);
+      }
+
+      var unboundMap = this.unboundMap_;
+      for ( var key in unboundMap ) {
+        if ( ! unboundMap.hasOwnProperty(key) ) continue;
```

Comment:
This and the loop above are identical other than the input map. Maybe make a method for doing the conversion?

Reply:
In order for `first` handling to be general, it would need to be input/output, so the helper method would return two things (the new string and the `first` state).

IMHO it's simpler like this (though repetitive). [If there were 3](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) or more loops, I would tend to agree: a method would be better.

Reply:
Or just return a string, and let the caller deal with mashing them together. The logic for joining two potentially-empty strings is pretty trivial.

Reply:
On second thought, I can use an output function to avoid requiring two return values. Updated logic with helper function is in `DetachedURLState.js`